db/transaction.go

```diff
@@ -0,0 +1,157 @@
+	"roci.dev/diff-server/util/time"
+)
+
+var (
+	// ErrClosed is the error returned from operations on a Transaction when
+	// it has already been closed.
+	ErrClosed = errors.New("Transaction is closed")
+)
+
+type Transaction struct {
+	db     *DB
+	head   Commit
+	me     *kv.MapEditor
+	wrote  bool
+	closed bool
+	name   string
+	args   types.List
+}
+
+// Closed returns true when the database has been closed. A database becomes
+// closed after Committ or Close is called.
+func (tx *Transaction) Closed() bool {
+	return tx.closed
+}
+
+// Get returns the JSON encoded value at the given id of the database.
+func (tx *Transaction) Get(id string) ([]byte, error) {
+	if tx.Closed() {
+		return nil, ErrClosed
+	}
+	value := tx.me.Get(types.String(id))
+	if value == nil {
+		return nil, nil
+	}
+	var b bytes.Buffer
+	err := nomsjson.ToJSON(value, &b)
+	return b.Bytes(), err
+}
+
+// Has returns true if the database has an entry with the given id.
+func (tx *Transaction) Has(id string) (bool, error) {
+	if tx.Closed() {
+		return false, ErrClosed
+	}
+	return tx.me.Has(types.String(id)), nil
+}
+
+// Scan returns a slice of ScanItems of the id-value pairs in the database. You
+// can use ScanOptions to get all the items with a certain prefix or limit the
+// number of results.
+func (tx *Transaction) Scan(opts ScanOptions) ([]ScanItem, error) {
+	if tx.Closed() {
+		return nil, ErrClosed
+	}
+	return scan(tx.me.Build().NomsMap(), opts)
+}
+
+// Put adds or updates an existing entry in the database.
+func (tx *Transaction) Put(id string, json []byte) error {
+	if tx.Closed() {
+		return ErrClosed
+	}
+
+	canonicalJSON, err := nomsjson.Canonicalize(json)
+	if err != nil {
+		return fmt.Errorf("could not Put '%s'='%s': %w", id, json, err)
+	}
+	value, err := nomsjson.FromJSON(bytes.NewReader(canonicalJSON), tx.db.noms)
+	if err != nil {
+		return fmt.Errorf("could not Put '%s'='%s': %w", id, json, err)
+	}
+
+	err = tx.me.Set(types.String(id), value)
+	if err != nil {
+		return fmt.Errorf("could not Put '%s'='%s': %w", id, value, err)
+	}
+
+	tx.wrote = true
+	return nil
+}
+
+// Del removes an entry from the database. It returns true if the entry existed
+// before the call to Del.
+func (tx *Transaction) Del(id string) (ok bool, err error) {
+	if tx.Closed() {
+		return false, ErrClosed
+	}
+
+	k := types.String(id)
+	ok = tx.me.Has(k)
+	if ok {
+		err = tx.me.Remove(k)
+		if err == nil {
+			tx.wrote = true
+		}
+	}
+	return ok, err
+}
+
+// Close the transaction without committing any possible changes done in this
+// transaction.
+func (tx *Transaction) Close() error {
+	if tx.closed {
+		return ErrClosed
+	}
+	tx.closed = true
+	return nil
+}
+
+// Commit tries to commits the changes made to the database in this transaction.
+// If this returns without an error the commit succeeded and the new ref of the
+// database head is returned.
+func (tx *Transaction) Commit() (ref types.Ref, err error) {
+	if tx.closed {
```

Comment:
i feel just a little discomfort that closed is not protected by a mutex. it looks like so long as you switch the order of marking a transaction closed/committed and removing it from the map (see comment elsewhere) we shouldn't ever call commit concurrently, however... might we feel a little safer if we knew it was *impossible* for a transaction to be committed twice (eg by adding a mutex on closed)?

Reply:
I agree that after guarding the maps it cannot happen.

Adding a mutex to ensure that this never cause any problem in the future

Reply:
Thinking more... I think we might want to guard every method of a tx with a mutex. For example we do not want a `Put` to happen in the middle of a `Commit`.

Reply:
yes, that's an even stronger condition than I suggested in https://github.com/rocicorp/replicache-client/pull/60#pullrequestreview-395086277 and i think it makes sense. we will need to be sure we are checking closed inside the lock in case someone else closed or committed while we were waiting.